### PR TITLE
SONARJAVA-1298 Type substitution in parametrized methods

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 1]
```diff
@@ -96,13 +96,16 @@ private JavaType substituteTypeParameter(JavaType type, TypeSubstitution substit
...
-    if (type instanceof JavaType.WildCardType) {
+    } else if (type instanceof JavaType.WildCardType) {
       JavaType.WildCardType wildCardType = (JavaType.WildCardType) type;
       substitutedType = substitution.substitutedType(wildCardType.bound);
       if (substitutedType != null) {
         return parametrizedTypeCache.getWildcardType(substitutedType, wildCardType.boundType);
       }
+    } else if (type instanceof JavaType.ArrayJavaType) {
+      JavaType elementType = ((JavaType.ArrayJavaType) type).elementType;
+      // FIXME SONARJAVA-1574
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "substituteTypeParameter" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/8876d66fe7fd51e266df5391dadee2441869d0ed/java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java#L87) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 2]
```diff
@@ -478,24 +477,11 @@ private Resolution findMethod(Env env, JavaType callSite, JavaType site, String
...
-    }
-    return type;
-  }
-
   /**
    * @param candidate    candidate
    * @param bestSoFar previously found best match
    */
-  private JavaSymbol selectBest(Env env, JavaType site, List<JavaType> argTypes, JavaSymbol candidate, JavaSymbol bestSoFar, boolean autoboxing) {
+  private JavaSymbol selectBest(Env env, JavaType site, List<JavaType> argTypes, List<JavaType> typeParams, JavaSymbol candidate, JavaSymbol bestSoFar, boolean autoboxing) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR The Cyclomatic Complexity of this method "selectBest" is 16 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 3]
```diff
@@ -604,13 +636,13 @@ private boolean isAcceptableByAutoboxing(JavaType expressionType, JavaType forma
   /**
    * JLS7 15.12.2.5. Choosing the Most Specific Method
    */
-  private JavaSymbol selectMostSpecific(JavaSymbol m1, JavaSymbol m2) {
+  private JavaSymbol selectMostSpecific(JavaSymbol m1, JavaSymbol m2, TypeSubstitution substitution) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR The Cyclomatic Complexity of this method "selectMostSpecific" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 4]
```diff
@@ -560,22 +476,21 @@ private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> fo
...
       JavaType argType = argTypes.get(argsSize - i);
       // check type of element of array or if we invoke with an array that it is a compatible array type
-      if (!isAcceptableType(argType, lastFormal.elementType, autoboxing, isParameterized) && (nbArgToCheck != 1 || !types.isSubtype(argType, lastFormal))) {
+      if (!isAcceptableType(argType, lastFormal.elementType, autoboxing) && (nbArgToCheck != 1 || !types.isSubtype(argType, lastFormal))) {
         return false;
       }
     }
     for (int i = 0; i < argsSize - nbArgToCheck; i++) {
-      if (!isAcceptableType(argTypes.get(i), formals.get(i), autoboxing, isParameterized)) {
+      if (!isAcceptableType(argTypes.get(i), formals.get(i), autoboxing)) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR The Cyclomatic Complexity of this method "selectBest" is 16 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "selectMostSpecific" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/065cbb18a2e9934b44e8b82605e9f7e6afb24d47/java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java#L639) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+  private JavaType applySubstitution(JavaType.ArrayJavaType arrayType, TypeSubstitution substitution) {
+    JavaType rootElementType = arrayType.elementType;
+    int nbDimensions = 1;
+    while (rootElementType.isTagged(JavaType.ARRAY)) {
+      rootElementType = ((JavaType.ArrayJavaType) rootElementType).elementType;
+      nbDimensions++;
+    }
+    JavaType substitutedType = substitution.substitutedType(rootElementType);
+    if (substitutedType != null) {
+      // FIXME SONARJAVA-1574 a new array type should not be created but reused if already existing for the current element type
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 6]
```diff
@@ -520,28 +453,11 @@ private JavaSymbol selectBest(Env env, JavaType site, List<JavaType> argTypes, J
...
-  private static boolean usesTypeParameter(JavaSymbol.MethodJavaSymbol methodSymbol) {
-    return !methodSymbol.typeVariableTypes.isEmpty();
-  }
-
   /**
    * @param argTypes types of arguments
    * @param formals  types of formal parameters of method
    */
-  private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> formals, boolean isVarArgs, boolean autoboxing, boolean isParameterized) {
+  private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> formals, boolean isVarArgs, boolean autoboxing) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR The Cyclomatic Complexity of this method "isArgumentsAcceptable" is 14 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by m-g-sonar** [comment]
Continued on https://github.com/SonarSource/sonar-java/pull/714

**Pull request closed without merging**